# HG changeset patch
# Parent fc60215fa2fa2e663371b2857d7ccba36320735f
# User Sukhbir Singh <sukhbir.in@gmail.com>
Added support for removing the Date header and setting a custom Message-ID.


diff --git a/mailnews/compose/src/nsMsgCompUtils.cpp b/mailnews/compose/src/nsMsgCompUtils.cpp
--- a/mailnews/compose/src/nsMsgCompUtils.cpp
+++ b/mailnews/compose/src/nsMsgCompUtils.cpp
@@ -364,36 +364,42 @@ mime_generate_headers (nsMsgCompFields *
     if (deliver_mode == MSG_SaveAsTemplate) {
       const char *pStr = fields->GetTemplateName();
       pStr = pStr ? pStr : "";
       ENCODE_AND_PUSH("X-Template: ", false, pStr, charset, usemime);
     }
 #endif /* SUPPORT_X_TEMPLATE_NAME */
   }
 
-  PRExplodedTime now;
-  PR_ExplodeTime(PR_Now(), PR_LocalTimeParameters, &now);
-  int gmtoffset = (now.tm_params.tp_gmt_offset + now.tm_params.tp_dst_offset) / 60;
+  bool remove_date = true;
+  prefs->GetBoolPref("mailnews.local_date_header_generation", &remove_date);
 
-  /* Use PR_FormatTimeUSEnglish() to format the date in US English format,
-     then figure out what our local GMT offset is, and append it (since
-     PR_FormatTimeUSEnglish() can't do that.) Generate four digit years as
-     per RFC 1123 (superceding RFC 822.)
-   */
-  PR_FormatTimeUSEnglish(buffer_tail, 100,
-               "Date: %a, %d %b %Y %H:%M:%S ",
-               &now);
+  if (remove_date)
+  {
+    PRExplodedTime now;
+    PR_ExplodeTime(PR_Now(), PR_LocalTimeParameters, &now);
+    int gmtoffset = (now.tm_params.tp_gmt_offset + now.tm_params.tp_dst_offset) / 60;
 
-  buffer_tail += PL_strlen (buffer_tail);
-  PR_snprintf(buffer_tail, buffer + size - buffer_tail,
-        "%c%02d%02d" CRLF,
-        (gmtoffset >= 0 ? '+' : '-'),
-        ((gmtoffset >= 0 ? gmtoffset : -gmtoffset) / 60),
-        ((gmtoffset >= 0 ? gmtoffset : -gmtoffset) % 60));
-  buffer_tail += PL_strlen (buffer_tail);
+    /* Use PR_FormatTimeUSEnglish() to format the date in US English format,
+       then figure out what our local GMT offset is, and append it (since
+       PR_FormatTimeUSEnglish() can't do that.) Generate four digit years as
+       per RFC 1123 (superceding RFC 822.)
+     */
+    PR_FormatTimeUSEnglish(buffer_tail, 100,
+                 "Date: %a, %d %b %Y %H:%M:%S ",
+                 &now);
+
+    buffer_tail += PL_strlen (buffer_tail);
+    PR_snprintf(buffer_tail, buffer + size - buffer_tail,
+          "%c%02d%02d" CRLF,
+          (gmtoffset >= 0 ? '+' : '-'),
+          ((gmtoffset >= 0 ? gmtoffset : -gmtoffset) / 60),
+          ((gmtoffset >= 0 ? gmtoffset : -gmtoffset) % 60));
+    buffer_tail += PL_strlen (buffer_tail);
+  }
 
   if (pFrom && *pFrom)
   {
     ENCODE_AND_PUSH("From: ", true, pFrom, charset, usemime);
   }
 
   if (pReplyTo && *pReplyTo)
   {
@@ -1110,18 +1116,36 @@ msg_generate_message_id (nsIMsgIdentity 
   }
 
   if (!isValidHost(host))
   /* If we couldn't find a valid host name to use, we can't generate a
      valid message ID, so bail, and let NNTP and SMTP generate them. */
     return 0;
 
   GenerateGlobalRandomBytes((unsigned char *) &salt, sizeof(salt));
-  return PR_smprintf("<%lX.%lX@%s>",
-           (unsigned long) now, (unsigned long) salt, host);
+
+  bool custom_message_id = false;
+  char *message_id = nsnull;
+
+  nsCOMPtr<nsIPrefBranch> prefs(do_GetService(NS_PREFSERVICE_CONTRACTID));
+
+  prefs->GetBoolPref("mailnews.custom_message_id", &custom_message_id);
+  prefs->GetCharPref("mailnews.header.custom_message_id", &message_id);
+
+  if (!custom_message_id)
+  { 
+    return PR_smprintf("<%lX.%lX@%s>",
+              (unsigned long) now, (unsigned long) salt, host);
+  } 
+  else 
+  {
+    return PR_smprintf("<%s.%lX@%s>",
+              message_id, (unsigned long) salt, host);
+  }
+
 }
 
 
 inline static bool is7bitCharset(const nsCString& charset)
 {
   // charset name is canonical (no worry about case-sensitivity)
   return charset.EqualsLiteral("HZ-GB-2312") ||
          Substring(charset, 0, 8).EqualsLiteral("ISO-2022-");
